### PR TITLE
Link CUDA libraries and executables with nvcc.

### PR DESCRIPTION
**Pull request by robinson96**
With host compilation, there is no way to link in a way that allows kernels in one library to call device functions defined in another library. Linking both the libraries and the executable with nvcc resolves this issue.

### CODE REVIEW

**Comment by davidbeckingsale** [comment]
How easy is it to trigger the breaking behavior? 

Perhaps it would be a good idea to add a test that would fail without this PR - that way if we ever change the FindCUDA stuff (e.g. using CMake's built-in CUDA "support") we won't get bit by this again.

**Comment by cyrush** [comment]
it would be good to create a test for this case, b/c if we refactor cuda stuff (for ex: using `cuda_wrap_srcs` instead of `cuda_add_library`), we want to make sure to not break this.

Guess we need to test cuda compilation in travis.

**Comment by cyrush** [comment]
@davidbeckingsale hah, we are on the same wavelength.

**Comment by robinson96** [comment]
I have a pretty small reproducer, not hooked up to BLT yet. Where / how should I set it up?

**Comment by davidbeckingsale** [comment]
There are some examples in the `tests` subdirectory.

**Comment by cyrush** [comment]
@robinson96 

You can add as part of the internal test project:
https://github.com/LLNL/blt/tree/master/cmake/blt-test/src

(even if cuda isn't tested in travis, you should be able to test locally and vouch until we enable cuda testing)

**Comment by cyrush** [comment]
@davidbeckingsale , we don't really want it as part of `tests` b/c those are the smoke tests we want users to run to identify env issues (vs tests of blt itself)

**Comment by davidbeckingsale** [comment]
@cyrush - you're right, I ninja-edited my comment. I guess the email notification was for the original version.

**Comment by robinson96** [comment]
The reproducer has 6 or so src files associated with it, so should probably be grouped in a separate directory. Do you want that under blt-test/src or alongside it?

**Comment by robinson96** [comment]
(and I should assume that things normally defined in a host config will be implemented when cuda testing is enabled, right?)

**Comment by cyrush** [comment]
@robinson96, 

If thats easiest for now, sounds good.

It seems like we could get it down to two libs + a test, which could eventually fit into the master test.

**Comment by robinson96** [comment]
Note that this issue is only actually reproduces if -G is in the NVCC_FLAGS.
Longer term, to have a robust test that doesn’t depend on proper host-configs, 
we might want setup and teardown routines that generate appropriate host-configs
for the test. 

I’ve modified the ray host config in the tutorial host-configs so it can be used
for this test. 

Note t_git_macros_smoke is failing - but it’s failing on master so it’s not my 
fault! (at least in this branch). 


Before changes:
Test project /g/g18/probinso/git/blt/cmake/blt-test/build
    Start 1: t_example_smoke
1/8 Test #1: t_example_smoke ..................   Passed    0.01 sec
    Start 2: t_header_only_smoke
2/8 Test #2: t_header_only_smoke ..............   Passed    0.01 sec
    Start 3: t_git_macros_smoke
3/8 Test #3: t_git_macros_smoke ...............***Failed    0.01 sec
    Start 4: t_device_call_from_kernel
4/8 Test #4: t_device_call_from_kernel ........***Failed    0.14 sec
    Start 5: blt_gtest_smoke
5/8 Test #5: blt_gtest_smoke ..................   Passed    0.01 sec
    Start 6: blt_mpi_smoke
6/8 Test #6: blt_mpi_smoke ....................   Passed    0.72 sec
    Start 7: blt_cuda_smoke
7/8 Test #7: blt_cuda_smoke ...................   Passed    0.12 sec
    Start 8: blt_cuda_runtime_smoke
8/8 Test #8: blt_cuda_runtime_smoke ...........   Passed    0.04 sec


After changes:
Test project /g/g18/probinso/git/blt/cmake/blt-test/build
    Start 1: t_example_smoke
1/8 Test #1: t_example_smoke ..................   Passed    0.01 sec
    Start 2: t_header_only_smoke
2/8 Test #2: t_header_only_smoke ..............   Passed    0.01 sec
    Start 3: t_git_macros_smoke
3/8 Test #3: t_git_macros_smoke ...............***Failed    0.01 sec
    Start 4: t_device_call_from_kernel
4/8 Test #4: t_device_call_from_kernel ........   Passed    0.12 sec
    Start 5: blt_gtest_smoke
5/8 Test #5: blt_gtest_smoke ..................   Passed    0.01 sec
    Start 6: blt_mpi_smoke
6/8 Test #6: blt_mpi_smoke ....................   Passed    0.66 sec
    Start 7: blt_cuda_smoke
7/8 Test #7: blt_cuda_smoke ...................   Passed    0.12 sec
    Start 8: blt_cuda_runtime_smoke
8/8 Test #8: blt_cuda_runtime_smoke ...........   Passed    0.04 sec

**Comment by cyrush** [comment]
Getting a linking error on my osx machine (CUDA 8, XCode 7.3):

(This is just w/ ENABLE_CUDA=ON, no other flags)

```
/usr/local/cuda/bin/nvcc /Users/harrison37/Work/blt_work/github/blt/cmake/blt-test/blt_device_call_from_kernel/Child.cxx -x=cu -c -o /Users/harrison37/Work/blt_work/github/blt/cmake/blt-test/build/CMakeFiles/t_device_call_from_kernel_lib.dir/blt_device_call_from_kernel/./t_device_call_from_kernel_lib_generated_Child.cxx.o -ccbin /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -m64 -DUSE_CUDA -DNVCC -I/usr/local/cuda/include
nvcc warning : The 'compute_20', 'sm_20', and 'sm_21' architectures are deprecated, and may be removed in a future release (Use -Wno-deprecated-gpu-targets to suppress warning).
/Users/harrison37/Work/blt_work/github/blt/cmake/blt-test/blt_device_call_from_kernel/Child.cxx(31): warning: variable "e" was set but never used

/Users/harrison37/Work/blt_work/github/blt/cmake/blt-test/blt_device_call_from_kernel/Child.cxx(31): warning: variable "e" was set but never used

ptxas fatal   : Unresolved extern function '_ZN6ParentC2EPKci'
```

Missing link to:
`Parent::Parent(char const*, int)`

**Comment by robinson96** [comment]
@cyrush : are you using a host-config in blt? check out the ray host-config in the docs/tutorials for the other knobs that need to be turned on...

**Comment by cyrush** [comment]
@robinson96 I did look at the ray host-config, but its not clear to me which options are necessary vs those specific to ray.  
 
```
set (CUDA_ARCH "sm_60" CACHE PATH "")
set (NVCC_FLAGS -restrict; -arch; ${CUDA_ARCH}; -std c++11; --expt-extended-lambda; -G )
```

For example, I can't use sm60 b/c my laptop doesn't have sm60.

Looking at the code, i don't really see how `-restrict` or `-std c++11` would make a difference for this demo?  Also, you stated the problem happens when `-G` is there, but I get linking issue even w/o `-G`. 

Just trying to make sure we are getting it right on a few platforms (esp since ray is essentially an experimental machine) 

(Side note: BLT should be passing -std c++11 to CUDA, but that doesn't appear to be the case)

**Comment by robinson96** [comment]
the -G is important so you don't get false passes. 
The likely cause of your error is that you need the following line
set (CUDA_SEPARABLE_COMPILATION ON CACHE BOOL "" )

**Comment by rhornung67** [comment]
FWIW, on blueos machines, I found I had to back off on the architecture for debug builds (code would die at runtime), and so I'm using -arch compute_30 for that. At least that is the case for CUDA 8. Haven't tried something different with CUDA 9, but I don't really care for debug builds.

**Comment by cyrush** [comment]
@robinson96 , well I am not getting a false pass :-), its simply not linking at all ...

**Comment by robinson96** [comment]
@cyrush , latest push should fix the Apple build.

**Comment by cyrush** [comment]
@robinson96  something more than meets the eye here.

I just tried the reproducer on ray using blt master (eg: w/o these changes) and the example links and runs for me. 

This was w/ cmake 3.9.2 w/ Cuda 8 and:

```
set(CUDA_NVCC_FLAGS " -G ")
set (CUDA_SEPARABLE_COMPILATION ON CACHE BOOL "" )
```

**Comment by robinson96** [comment]
@cyrush : Maybe the issue as Rich suggested is the combination of -G and the -arch sm_60.  
On reflection, I'm recalling another issue I encountered that cropped up when we moved from Keplers to Pascals that could have been a reproducer for this.  I think I agree that this is compensating for a bug on NVIDIAs end.  
That said - I'm not sure that I see a downside to nvcc linking instead of host compiler linking.  And, if it provides a smoother experience for BLT users on some platform, shouldn't that be considered a win?
Re Separable compilation and #ifdefs - I think those are orthogonal.  The #ifdefs just allow compilation of the test without NVCC.  Separable compilation is required for situations where the final link might not have all necessary information, as I believe is the case once you have kernels calling device methods in other libraries.

**Comment by maxpkatz** [comment]
@rhornung67 have you filed the -G and sm_60 issue as a bug with NVIDIA? If not, can you provide me a reproducer?

**Comment by robinson96** [comment]
Another tidbit - Danny Taller has tried dropping the sm_60 in the more real world example, and that does not appear to solve the problem, where as linking with nvcc does. Curioser and Curioser.

**Comment by cyrush** [comment]
@robinson96 

separable compilation stuff - got it. 

I checked again,  on ray -- I don't get a linking error but i do get a runtime error. 
`GPUassert: an illegal instruction was encountered /g/g24/cyrush/work/2017.10.blt-cuda-thingy/blt-test/blt_device_call_from_kernel/CudaTests.cxx 28
`
It does only happens when `-G"`is on. In this case sm_60 was not specified. 
I suspect thats what you experienced?

This does not happen on surface (chaos5), or on OSX when using `-G`, 

I am little hesitant to swing linking over to nvcc b/c:

1) It seems to be confined to blueos.

I really want to know what the root cause is to feel comfortable with the solution.
 
2)  I am not sure we have done a complete job at the surgery on FindCUDA to harden linking with nvcc.

Given that a mac issue popped right away & we haven't tested windows -- I fear we may end up in a wack-a-mole situation. 

It could smooth things out on blueos for folks, but also may cause headaches elsewhere. 

3) I haven't seen other projects need to force linking with nvcc.

**Comment by cyrush** [comment]
yes, I don't think sm_60 is the issue, it seems to be `-G` on blueos

**Comment by maxpkatz** [comment]
Issues with -G are orthogonal to the underlying linker issue. When you are trying to call a device function in a kernel, and the device function is part of a library, nvcc must be involved in some stage of the linking so that the kernel knows where to look to find the location of the device function. The host compiler does not know how to do this. This does not mean nvcc must be the only linker at the final build stage, but it does mean that nvcc needs to be involved, at least to do the device linking (see an example [here](http://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/index.html#nvcc-options-for-separate-compilation) for how to do separate device linking, with nvcc --device-link, and final linking with the host compiler).

**Comment by cyrush** [comment]
thanks for the context @maxpkatz.

We are using (slightly modified) CMake FindCUDA logic. 

With separable compilation on (and maybe even w/o) it breaks the build into two stages:

- Building NVCC (Device) object
- Building NVCC intermediate link file

I believe the second step is used to contain the nvcc linking device linking, but not 100%.


For our current issue, we don't actually hit any linking errors -- but we get a runtime error (illegal instruction). Is that the expected failure mode for a linking issue?

**Comment by DannyTallerLLNL** [comment]
@cyrush, yes, that is the issue I often run into. An illegal instruction error when a device function is invoked ( but before we enter the function and start doing work inside ). Curiously, I have found that if you re-order things in the code, you can sometimes get things to run to completion with correct results at runtime, but the problem still shows up in cuda-memcheck .

**Comment by maxpkatz** [comment]
I've attached the standalone version of the test case based on Tabale, that I think @cyrush and @robinson96 are referring to the modified version of. You can indeed "successfully" link when using the host compiler to link, but that does not mean it will run correctly. The host compiler doesn't even know that it should be attempting to do the correct relocations for the device code, so it will not complain. Note that if you remove the -G flag from this test, cuda-memcheck will not report any errors, but that does not mean the code actually works as intended -- if you insert a printf into Child::Evaluate, you'll see that the device function doesn't actually get called when using the incorrect linking strategy (using "make bad").

[test.tar.gz](https://github.com/LLNL/blt/files/1392545/test.tar.gz)

@cyrush, the strategy you mentioned is fine, but Danny's problem comes when you try to call one NVCC intermediate link file from another, which requires NVCC device linking between both of them to work.

**Comment by maxpkatz** [comment]
One thing is worth clarifying, based on my consultation with the local linker expert at NVIDIA. It is fine* to create a library using ar and ranlib; indeed, nvcc --lib does something similar. The "bad" case fails because of the device link before the library is created. So if you remove that intermediate link step, and only do the final link when you're joining everything together, that is the key. I've attached a modified version of the above test with a "good2" target that demonstrates this in a way that allows you to keep the final link being done by the host compiler.

[test.tar.gz](https://github.com/LLNL/blt/files/1392975/test.tar.gz)

Of course, that solves the problem with @DannyTallerLLNL's original example, not necessarily the BLT system. But I'm hoping you can make the appropriate changes based on this example.

*He also points out that ar rc is better than ar qc in this case, because you can run into build problems if the code gets appended multiple times on separate build invocations.

**Comment by cyrush** [comment]
@robinson96, I pushed my changes that add comments around the BLT FindCuda changes to make them easier to identify. You can search for `BEGIN BLT CHANGE`. If we support this via some option (which I think is a good idea), these would be places to insert those checks as well.

File: cmake/thirdparty/FindCUDA.cmake [hunk 1]
```diff
@@ -534,6 +552,13 @@ mark_as_advanced(
   CUDA_BUILD_EMULATION
   CUDA_VERBOSE_BUILD
   CUDA_SEPARABLE_COMPILATION
+  ####################
```
Comment (kennyweiss) [inline cmake/thirdparty/FindCUDA.cmake]: These ``{BEGIN/END} BLT CHANGE`` comments should probably be removed before merging.

Comment (cyrush) [reply cmake/thirdparty/FindCUDA.cmake]: we want those in there to make it very obvious what we changed w/o relying on git forensics

Comment (kennyweiss) [reply cmake/thirdparty/FindCUDA.cmake]: Thanks @cyrush -- didn't notice that it was an edited version of cmake's ``FindCUDA``

**Comment by jonesholger** [comment]
Necroed:  This is just a supposition that I absolutely need host link for final step (as when mixing cuda and omp-target). Assuming all of the objects were put into archives like in Max's example: You could extract them again, toss the device linked objects,  then recombine with a device link before the final host link step. This isn't really anything much different from Max's good2 example. But, I wanted to work through the steps

 goodcomb:
	$(CUDA_PATH)/bin/nvcc $(NVCC_COMP_FLAGS) Child.cxx -o Child.o
	$(CUDA_PATH)/bin/nvcc $(NVCC_COMP_FLAGS) Parent.cxx -o Parent.o
	$(CUDA_PATH)/bin/nvcc $(NVCC_LINK_FLAGS) Child.o Parent.o -o testprob_static_intermediate_link.o
	ar qc libpart1.a Child.o Parent.o testprob_static_intermediate_link.o
	ranlib libpart1.a
	$(CUDA_PATH)/bin/nvcc $(NVCC_COMP_FLAGS) CudaTests.cxx -o testExec.o
	$(CUDA_PATH)/bin/nvcc $(NVCC_LINK_FLAGS) testExec.o -o testExec_intermediate_link.o
	ar qc libpart2.a testExec.o testExec_intermediate_link.o
	ranlib libpart2.a 
	rm *.o
	ar x libpart1.a
	ar x libpart2.a
	$(CUDA_PATH)/bin/nvcc $(NVCC_LINK_FLAGS) Child.o Parent.o testExec.o -o testExec_intermediate_link_comb.o
	g++ -g Child.o Parent.o testExec.o -o testExec $(CUDA_PATH)/lib64/libcudart_static.a -ldl  testExec_intermediate_link_comb.o -lcuda -lpthread -lrt $(CUDA_PATH)/lib64/libcudart_static.a -ldl $(CUDA_PATH)/lib64/libcudadevrt.a

**Pull request merged**